akka-projection-core/src/main/scala/akka/projection/internal/InternalProjectionState.scala

```diff
@@ -256,14 +269,18 @@ private[akka] abstract class InternalProjectionState[Offset, Envelope](
     handlerStrategy match {
       case SingleHandlerStrategy(handler) =>
         source
-          .mapAsync(parallelism = 1) {
-            case (offset, envelope) =>
-              val processed =
-                saveOffset(projectionId, offset).flatMap { _ =>
-                  handlerRecovery
-                    .applyRecovery(envelope, offset, offset, abort.future, () => handler.process(envelope))
-                }
-              reportProgress(processed, envelope)
+          .mapAsync(parallelism = 1) { context =>
+            val processed =
+              saveOffset(projectionId, context.offset).flatMap { _ =>
+                handlerRecovery
+                  .applyRecovery(
+                    context.envelope,
+                    context.offset,
+                    context.offset,
```

Comment:
Seems like we could just pass the whole context to the recovery handler impl.

Reply:
I think we pass in different offsets in some cases, which wouldn't correspond to 1 context. firstOffset, lastOffset for the grouped case

Reply:
You're right, my mistake.

Reply:
Hmm, I think we can do it with the first two arguments. All usages of `applyRecovery()` use an envelope and an offset from the same context as first and second arguments. The third argument (aka, the second offset) may be from a different context. 👍

Reply:
I'd still do it on a separate PR, though.